### DISCUSSION TITLE
自己定义一个类，里面包含TcpClient，在kill 进程号退出时，没有正常退出

### DISCUSSION THREAD

**Discussion by wangdongyu1989** [chenshuo/muduo#603] [General]
报的日志：muduo::net::Channel::~Channel(): Assertion !addedToLoop_' failed.
connectDestroyed没有调用，没有执行channel::remove
然后channel再析构的时候，assert报错
想咨询下，什么情况下会出现这种问题以及如何解决？多谢多谢

**Comment by tangshiixang** [comment]
你的程序问题，是因为动作(channel::remove,为了线程安全)放在EventLoop里去跑，但是EventLoop被quit了，导致channel::remove（还在EventLoop中的pendingFunctors_中）未执行，同时TcpConnection在另一个线程析构，进而导致channel析构，引起你说的问题了。
无法解决，除非你能保证EventLoop在程序退出时结束循环，并且不关心程序退出时，一些资源释放动作未被执行(因为可能还在EventLoop中的pendingFunctors_中)。只要有runInLoop/queueInLoop这样的概念, 程序是多线程的情况下，还是会遇到你说看到的问题。EventLoop的quit就是会导致部分动作可能执行不了，而你程序为了线程安全又非得让它在EventLoop执行，这是矛盾的。
我猜测改库本来设计，就是不考虑EventLoop会在程序不退出情况下提前quit的。同时程序也不关心退出时，资源可能不能释放干净的问题。因为该库主要是用于局域网服务端程序的。
